diagnostics/diag.go

```diff
@@ -208,6 +211,18 @@ func (d *Diagnostics) getDiagnosticFromPeer(ctx context.Context, p peer.ID, pmes
 		return nil, err
 	}
 
+	// TEMP: trigger the multistream handshake with a read so we can
+	// use the old protocol name if needed
+	// TODO: remove once the network has switched over to the new code ( >= ipfs 0.4.2)
+	_, err = s.Read(nil)
+	if err != nil {
+		if strings.Contains(err.Error(), "protocol mismatch") {
+			s, err = d.host.NewStream(ctx, ProtocolDiagOld, p)
+		} else {
+			return nil, err
+		}
+	}
```

Comment:
~~i dont understand why this is needed? adding more protocols should be the same as using ...~~ ohhh i get it, if our first connecting fails, we issue a read to check if what we sent is bad-- hence use a different one? Wonder if paying to cost of 1RTT is better or worse than opening 2 streams simultaneously and closing one afterward.... hmmmmmmmmm decisions.